Change type of FrameLoader opener and ScriptController globalObject

pascoej opened the pull request:
#### b6dd5136a4bb992aa4e1a10923f03a36e8486d24
<pre>
Change type of FrameLoader opener and ScriptController globalObject
<a href="https://bugs.webkit.org/show_bug.cgi?id=255743">https://bugs.webkit.org/show_bug.cgi?id=255743</a>
rdar://108333335

Reviewed by Chris Dumez.

This patch does some refactoring required for PSON in windows opened
via window.open. The first part is changing the type of opener in
FrameLoader to accept a Frame instead of a LocalFrame. The second part
is changing the type of ScriptController&apos;s globalObject method to return
a global object instead of a local window.

* Source/WebCore/bindings/js/JSDOMWindowBase.cpp:
(WebCore::toJSLocalDOMWindow):
* Source/WebCore/bindings/js/ScriptController.h:
(WebCore::ScriptController::globalObject):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::fallbackBaseURL const):
(WebCore::Document::canNavigateInternal):
(WebCore::Document::initSecurityContext):
(WebCore::Document::initContentSecurityPolicy):
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::doCrossOriginOpenerHandlingOfResponse):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::detachFromAllOpenedFrames):
(WebCore::FrameLoader::opener):
(WebCore::FrameLoader::opener const):
(WebCore::FrameLoader::setOpener):
(WebCore::FrameLoader::setOriginalURLForDownloadRequest):
(WebCore::FrameLoader::updateRequestAndAddExtraFields):
(WebCore::FrameLoader::effectiveReferrerPolicy const):
* Source/WebCore/loader/FrameLoader.h:
* Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
(WebKit::WebLoaderStrategy::scheduleLoadFromNetworkProcess):
* Source/WebKit/WebProcess/WebCoreSupport/WebResourceLoadObserver.cpp:
(WebKit::WebResourceLoadObserver::logUserInteractionWithReducedTimeResolution):
* Source/WebKitLegacy/mac/WebView/WebFrame.mm:
(-[WebFrame _stringByEvaluatingJavaScriptFromString:withGlobalObject:inScriptWorld:]):

Canonical link: <a href="https://commits.webkit.org/263375@main">https://commits.webkit.org/263375@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/60756f21dc62149a3b9d0c9c1f9725ca74d284a5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/3922 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/4014 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/4125 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/5360 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/4182 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/3899 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/4099 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/4004 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/3995 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/3970 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/4165 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/3536 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/5229 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/1668 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/3511 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/5561 "") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/3488 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/3987 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/4996 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3976 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/3208 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/3495 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/3511 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1088 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/3543 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/3774 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKitLegacy/mac/WebView/WebFrame.mm:
```diff
@@ -2056,7 +2056,7 @@ - (NSString *)_stringByEvaluatingJavaScriptFromString:(NSString *)string withGlo
 
     // Get the frame frome the global object we've settled on.
-    auto* frame = anyWorldGlobalObject->wrapped().frame();
+    auto* frame = static_cast<const WebCore::JSLocalDOMWindow*>(anyWorldGlobalObject)->wrapped().frame();
```
Shouldn't this be a jsCast?

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/60756f21dc62149a3b9d0c9c1f9725ca74d284a5. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/12989

pascoej committed:
Change type of FrameLoader opener and ScriptController globalObject
https://bugs.webkit.org/show_bug.cgi?id=255743
rdar://108333335

Reviewed by Chris Dumez.

This patch does some refactoring required for PSON in windows opened
via window.open. The first part is changing the type of opener in
FrameLoader to accept a Frame instead of a LocalFrame. The second part
is changing the type of ScriptController's globalObject method to return
a global object instead of a local window.

* Source/WebCore/bindings/js/JSDOMWindowBase.cpp:
(WebCore::toJSLocalDOMWindow):
* Source/WebCore/bindings/js/ScriptController.h:
(WebCore::ScriptController::globalObject):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::fallbackBaseURL const):
(WebCore::Document::canNavigateInternal):
(WebCore::Document::initSecurityContext):
(WebCore::Document::initContentSecurityPolicy):
* Source/WebCore/loader/DocumentLoader.cpp:
(WebCore::DocumentLoader::doCrossOriginOpenerHandlingOfResponse):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::detachFromAllOpenedFrames):
(WebCore::FrameLoader::opener):
(WebCore::FrameLoader::opener const):
(WebCore::FrameLoader::setOpener):
(WebCore::FrameLoader::setOriginalURLForDownloadRequest):
(WebCore::FrameLoader::updateRequestAndAddExtraFields):
(WebCore::FrameLoader::effectiveReferrerPolicy const):
* Source/WebCore/loader/FrameLoader.h:
* Source/WebKit/WebProcess/Network/WebLoaderStrategy.cpp:
(WebKit::WebLoaderStrategy::scheduleLoadFromNetworkProcess):
* Source/WebKit/WebProcess/WebCoreSupport/WebResourceLoadObserver.cpp:
(WebKit::WebResourceLoadObserver::logUserInteractionWithReducedTimeResolution):
* Source/WebKitLegacy/mac/WebView/WebFrame.mm:
(-[WebFrame _stringByEvaluatingJavaScriptFromString:withGlobalObject:inScriptWorld:]):

Canonical link: https://commits.webkit.org/263375@main

webkit-commit-queue commented:
Committed 263375@main (b6dd5136a4bb): <https://commits.webkit.org/263375@main>

Reviewed commits have been landed. Closing PR #12989 and removing active labels.
